game_event/views.py

```diff
@@ -150,3 +154,32 @@ def process_answer_game_event(request, id):
     content['event_exists'] = True
 
     return render(request, 'game_event/result.html', content)
+
+
+def get_weather_from_api(game_events, weather_getter):
+    for game in game_events:
+        url = 'https://api.openweathermap.org/data/2.5/weather'
+        params = {
+            'q': game.court.city,
+            'units': 'metric',
+            'appid': os.getenv("APPID")
+        }
+
+        full_url = url + '?' + urllib.parse.urlencode(params)
+        try:
+            response = weather_getter(full_url)
+        except Exception:
+            game.weather = None
+            continue
+
+        if response.status == 200:
+            data = response.read().decode('utf-8')
+            weather_data = json.loads(data)
+            curr_icon = weather_data['weather'][0]['icon']
+            icon_url = "http://openweathermap.org/img/wn/" + curr_icon + "@2x.png"
+            curr_temp = weather_data['main']['temp']
+            game.weather = {'icon': icon_url, 'temp': curr_temp}
+        else:
+            game.weather = None
+
+    return game_events
```

Comment:
No need to return game_events since the method modifies the input list directly.

Reply:
It's better to return it. Functions that rely on side effects on the inputs are a mess.

Reply:
> It's better to return it. Functions that rely on side effects on the inputs are a mess.

I hadn't thought about this, but doesn't it still rely on side effects since we're updating the input variable `game_event`?
if I understand correctly function rely on side effect when it changes something outside of its function definition so to make this function pure we will need to create `new_game_events `list which copies the `game_events `and then make the changes on `new_game_events`, or am i missing something @yftacherzog ?

Reply:
It's cleanest if you have an  immutable object and instead of mutating it, you return a new object.

The current approach is not ideal, but I find the code easier to handle if you return the things you mutated rather than change them only as a side effect.